client/my-sites/controller.js

```diff
@@ -235,44 +246,78 @@ module.exports = {
+		const waitForInProgressSitesFetch = () => {
+			return new Promise( ( resolve, reject ) => {
+				if ( ! sites.fetched && sites.fetching ) {
+					sites.once( 'change', () => sites.fetched ? resolve() : reject() );
+				} else {
+					resolve();
+				}
+			} );
+		};
 
-			// if there was a redirect, we should terminate processing of next routes
-			// and let the redirect proceed
-			if ( ! selectionComplete ) {
-				return;
+		const hardFetchSites = () => new Promise( ( resolve, reject ) => {
+			sites.initialized = false;
+			sites.once( 'change', () => {
+				if ( sites.fetched ) {
+					sites.initialized = true;
+					return resolve();
+				}
+
+				reject();
+			} );
+
+			sites.fetch();
+		} );
+
+		let waitingNotice;
+		const validateAllVisibileSitesAreFetched = () => new Promise( resolve => {
+			if ( currentUser.visible_site_count !== sites.getVisible().length ) {
+				if ( ! waitingNotice ) {
+					waitingNotice = notices.info( i18n.translate( 'Finishing set up…' ), { showDismiss: false } );
+				}
+
+				return hardFetchSites().then(
+					validateAllVisibileSitesAreFetched,
+					validateAllVisibileSitesAreFetched
+				);
 			}
-		} else {
-			// if sites has fresh data and siteID is invalid
-			// redirect to allSitesPath
-			if ( sites.fetched || ! sites.fetching ) {
-				return page.redirect( allSitesPath );
+
+			if ( waitingNotice ) {
+				notices.removeNotice( waitingNotice );
 			}
 
-			let waitingNotice;
-			const selectOnSitesChange = () => {
-				// if sites have loaded, but siteID is invalid, redirect to allSitesPath
-				if ( sites.select( siteID ) ) {
-					sites.initialized = true;
-					onSelectedSiteAvailable( context );
-					if ( waitingNotice ) {
-						notices.removeNotice( waitingNotice );
-					}
-				} else if ( ( currentUser.visible_site_count !== sites.getVisible().length ) ) {
-					sites.initialized = false;
-					waitingNotice = notices.info( i18n.translate( 'Finishing set up…' ), { showDismiss: false } );
-					sites.once( 'change', selectOnSitesChange );
-					sites.fetch();
-				} else {
-					page.redirect( allSitesPath );
+			resolve();
+		} );
+
+		// ensure we have fetched sites
+		const sitesFetchedPromise = waitForInProgressSitesFetch().then( validateAllVisibileSitesAreFetched );
+
+		// Select a site
+		sitesFetchedPromise
+			.catch( error => {
+				page.redirect( allSitesPath );
+				return Promise.reject( error );
+			} )
+			.then( () => {
+				if ( ! sites.select( siteID ) ) {
+					return page.redirect( allSitesPath );
 				}
-			};
-			// Otherwise, check when sites has loaded
-			sites.once( 'change', selectOnSitesChange );
-		}
+
+				const selectionContext = {
+					reactContext: context,
+					selectedSite: sites.getSelectedSite()
+				};
+
+				feedReduxStoreWithSelectedSite( selectionContext );
+
+				if ( isSelectedSiteDomainOnlySite( context.store.getState() ) && ! isPathAllowedForDomainOnlySite( selectionContext ) ) {
```

Comment:
IMO, yes, selectors select data from redux stores, `sites` is not a redux store, I prefer to get whatever possible from redux store, so I prefer having a single selector that operates on a single store instead of 1 selector and 1 method that operates on different stores.